beacon-chain/execution/engine_client.go

```diff
@@ -315,6 +319,43 @@ func (s *Service) ExecutionBlockByHash(ctx context.Context, hash common.Hash, wi
 	return result, handleRPCError(err)
 }
 
+// ExecutionBlockByHashes fetches a batch of execution engine blocks by hash by calling
+// eth_blockByHash via JSON-RPC.
+func (s *Service) ExecutionBlockByHashes(ctx context.Context, hashes []common.Hash, withTxs bool) ([]*pb.ExecutionBlock, error) {
+	ctx, span := trace.StartSpan(ctx, "powchain.engine-api-client.ExecutionBlockByHashes")
+	defer span.End()
+	numOfHashes := len(hashes)
+	elems := make([]gethRPC.BatchElem, 0, numOfHashes)
+	execBlks := make([]*pb.ExecutionBlock, 0, numOfHashes)
+	errs := make([]error, 0, numOfHashes)
+	if numOfHashes == 0 {
+		return execBlks, nil
+	}
+	for _, h := range hashes {
+		blk := &pb.ExecutionBlock{}
+		err := error(nil)
+		newH := h
+		elems = append(elems, gethRPC.BatchElem{
+			Method: ExecutionBlockByHashMethod,
+			Args:   []interface{}{newH, withTxs},
+			Result: blk,
+			Error:  err,
+		})
+		execBlks = append(execBlks, blk)
+		errs = append(errs, err)
+	}
+	ioErr := s.rpcClient.BatchCall(elems)
+	if ioErr != nil {
+		return nil, ioErr
+	}
+	for _, e := range errs {
+		if e != nil {
+			return nil, handleRPCError(e)
```

Comment:
If one of the requests fail, do we fail the whole thing?

Reply:
Yeap, we do as the whole batch now becomes inconsistent.